core/src/main/java/org/opensearch/sql/calcite/utils/binning/BinFieldValidator.java

```diff
@@ -59,4 +60,39 @@ public static boolean isTimeBasedField(RelDataType fieldType) {
     // Check if type string contains EXPR_TIMESTAMP
     return fieldType.toString().contains("EXPR_TIMESTAMP");
   }
+
+  /** Validates that the field type is numeric for numeric binning operations. */
+  public static void validateNumericField(RelDataType fieldType, String fieldName) {
+    if (!isNumericField(fieldType)) {
+      throw new SemanticCheckException(
+          String.format(
+              "Cannot apply binning: field '%s' is non-numeric and not time-related, expected"
+                  + " numeric or time-related type",
+              fieldName));
+    }
+  }
+
+  /** Checks if the field type is numeric. */
+  private static boolean isNumericField(RelDataType fieldType) {
```

Comment:
thought: This is a pretty useful general utility outside of `bin`

Should we define it closer to where `SqlTypeName` is defined? We have lots of hand-rolled field logic for different methods and it's led to some issues when the type list change (e.g. OpenSearch recently introducing `TIMESTAMP_MICROS`).